### PR TITLE
feat: add delimiterless multi-item scoring (MaskMode::kMultiItemScoringV2)

### PR DESCRIPTION
**Pull request by chanh**
## 📌 Description

Adds `MaskMode::kMultiItemScoringV2` (value=4), a new attention mask mode that eliminates delimiter tokens from multi-item scoring sequences.

**MIS v1 (existing):** sequences include a delimiter token before each item; the kernel uses `token_pos_in_items` + `max_item_len_ptr` to reconstruct the mask.

**MIS v2 (this PR):** no delimiter tokens — item boundaries are passed via `item_offsets` (CSR-style array). The kernel precomputes a per-token `item_start` index for O(1) mask lookup.

Also fixes two bugs in the existing MIS v1 Python path:
- Removes shadowed `import torch` inside `if item_offsets` blocks that caused `UnboundLocalError` on the v1 code path
- Casts `uint32` tensors to `int64` before comparison (`offsets_b > 0`) since PyTorch does not support `compare_cuda` for `uint32`

### What MIS v2 saves

Removing delimiter tokens reduces the **total sequence length** through the entire transformer stack. With N items, v2 saves N tokens per layer (attention + MLP + layernorm). The savings are most significant when the delimiter-to-item-token ratio is high (many short items).

### Attention kernel benchmark (H100 80GB, FA2 backend)

Both V1 and V2 exclude prefix from Q — the only difference is delimiter tokens:

#### batch_size=1

| Config | V1 qo | V2 qo | Delims | V1 (ms) | V2 (ms) | Speedup | Delim% |
|--------|-------|-------|--------|---------|---------|---------|--------|
| 256 prefix, 100×3 items | 400 | 300 | 100 | 0.085 | 0.068 | **1.25×** | 25% |
| 256 prefix, 500×3 items | 2000 | 1500 | 500 | 0.081 | 0.078 | **1.03×** | 25% |
| 512 prefix, 200×5 items | 1200 | 1000 | 200 | 0.083 | 0.068 | **1.22×** | 17% |
| 256 prefix, 100×10 items | 1100 | 1000 | 100 | 0.064 | 0.054 | **1.18×** | 9% |
| 4096 prefix, 4×512 items | 2052 | 2048 | 4 | 0.498 | 0.539 | 0.92× | 0.2% |

#### batch_size=10

| Config | V1 qo | V2 qo | Delims | V1 (ms) | V2 (ms) | Speedup | Delim% |
|--------|-------|-------|--------|---------|---------|---------|--------|
| 256 prefix, 500×3 items | 2000 | 1500 | 500 | 0.465 | 0.396 | **1.17×** | 25% |
| 512 prefix, 200×5 items | 1200 | 1000 | 200 | 0.431 | 0.409 | **1.05×** | 17% |
| 4096 prefix, 4×512 items | 2052 | 2048 | 4 | 4.635 | 5.044 | 0.92× | 0.2% |

**Key takeaways:**
- Attention kernel speedup from removing delimiters is **modest** (1.0–1.25× at high delimiter ratios)
- At low delimiter ratios (<2%), v2 is slightly slower than v1 due to different mask logic
- The **primary benefit** is reducing total sequence length through the full transformer stack (attention + MLP + layernorm), which compounds across layers — this is **not measured** by the attention-only benchmark
- Best use case: many short items (e.g., 100+ items of 3-10 tokens each)

Benchmark script: `benchmarks/bench_mis_v2.py`

## 🔍 Related Issues

N/A

### ✅ Pre-commit Checks

- [ ] I have installed `pre-commit` by running `pip install pre-commit` (or used your preferred method).
- [ ] I have installed the hooks with `pre-commit install`.
- [ ] I have run the hooks manually with `pre-commit run --all-files` and fixed any reported issues.

## 🧪 Tests

Three test functions added to `tests/attention/test_batch_prefill_kernels.py`, all verified passing on GPU (H100, SM90a, CUDA 13.0, torch 2.11.0+cu130):

- **`test_batch_prefill_with_paged_kv_cache_multi_item_scoring_v2`** — V2 output matches explicit custom boolean mask reference (FA2 + Hopper paths)
- **`test_mis_v2_single_item_equals_causal`** — V2 with a single item equals standard causal `single_prefill_with_kv_cache` (unambiguous ground truth; no special mask logic in reference)
- **`test_mis_v1_v2_output_parity`** — V2 item-token outputs match V1 item-token outputs for equivalent multi-item inputs (shared Q/K/V at item and prefix positions; delimiter K/V in V1 is random but masked to -inf)

All 28 test cases pass.

**Files changed:**
- `include/flashinfer/attention/mask.cuh` — `kMultiItemScoringV2` enum value
- `include/flashinfer/attention/prefill.cuh` — FA2 kernel MIS v2 mask logic
- `include/flashinfer/attention/default_prefill_params.cuh` — `item_offsets`/`item_offsets_len` params
- `include/flashinfer/attention/hopper/mainloop_mma.cuh`, `prefill_sm90.cuh`, `sparse_mainloop.cuh`, `default_params.cuh` — Hopper kernel support
- `flashinfer/prefill.py`, `flashinfer/utils.py`, `flashinfer/jit/utils.py`, `flashinfer/jit/attention/modules.py` — Python API wiring
- `benchmarks/bench_mis_v2.py` — Fair V1/V2 attention kernel benchmark
- `benchmarks/debug_mis_speedup.py` — Diagnostic script isolating speedup sources

## Reviewer Notes

The MIS v2 mask is semantically equivalent to MIS v1 but without delimiter tokens. The single-item causal test provides ground-truth validation independent of V1. The V1/V2 parity test cross-validates multi-item behavior.

Validation guards reject `item_offsets` for unsupported backends (`cudnn`, `trtllm-gen`, `cutlass`) and for `use_cuda_graph=True` (MIS v2 allocates fresh tensors on each `plan()` call).

### CODE REVIEW

File: flashinfer/prefill.py [hunk 1]
```diff
@@ -1710,6 +1728,46 @@ def plan(
         self._token_pos_in_items_len = token_pos_in_items_len
         self._max_item_len_ptr = max_item_len_ptr
 
+        # V2 MIS: precompute item_start from item_offsets
+        self._item_start_ptr = None
+        self._item_start_len = 0
+        if item_offsets is not None:
+            batch_size = len(prefix_len_ptr) if prefix_len_ptr is not None else 1
```
[the remark this answers is not in the document]
Comment (chanh) [reply flashinfer/prefill.py]: Fixed — extracted the precomputation into a shared _precompute_item_start() helper that takes batch_size from qo_indptr (the authoritative source). Both the ragged and paged paths now call this helper with the correct batch_size.

File: flashinfer/prefill.py [hunk 2]
```diff
@@ -1710,6 +1728,46 @@ def plan(
...
+            item_start_list = []
+            max_items_tokens = 0
+            for b in range(batch_size):
+                offsets_b = item_offsets[b * item_offsets_len:(b + 1) * item_offsets_len]
+                # Filter out padding (zeros after the valid offsets)
+                num_offsets = (offsets_b.to(torch.int64) > 0).sum().item() + 1  # +1 for leading 0
+                valid_offsets = offsets_b[:num_offsets].long()
```
[the remark this answers is not in the document]
Comment (chanh) [reply flashinfer/prefill.py]: Fixed — now uses `nonzero()[-1]` to find the last non-zero index instead of counting `> 0`. This correctly handles cases like `offsets = [0, 0, 3]`. Also vectorized the inner loop with `torch.repeat_interleave` to eliminate Python-side `.item()` calls.

File: flashinfer/prefill.py [hunk 3]
```diff
@@ -2705,6 +2770,46 @@ def plan(
         self._token_pos_in_items_len = token_pos_in_items_len
         self._max_item_len_ptr = max_item_len_ptr
 
+        # V2 MIS: precompute item_start from item_offsets
+        self._item_start_ptr = None
+        self._item_start_len = 0
+        if item_offsets is not None:
+            batch_size_v2 = len(prefix_len_ptr) if prefix_len_ptr is not None else 1
```
[the remark this answers is not in the document]
Comment (chanh) [reply flashinfer/prefill.py]: Fixed — same as the ragged path: both now use the shared `_precompute_item_start()` helper which receives `batch_size` from `len(qo_indptr) - 1`.

File: flashinfer/prefill.py [hunk 4]
```diff
@@ -1710,6 +1728,46 @@ def plan(
...
+                valid_offsets = offsets_b[:num_offsets].long()
+                # Build per-token item_start
+                total_tokens = valid_offsets[-1].item()
+                max_items_tokens = max(max_items_tokens, total_tokens)
+                item_start_b = torch.zeros(total_tokens, dtype=torch.uint32, device=item_offsets.device)
+                for i in range(num_offsets - 1):
+                    start = valid_offsets[i].item()
+                    end = valid_offsets[i + 1].item()
+                    item_start_b[start:end] = start
+                item_start_list.append(item_start_b)
```
[the remark this answers is not in the document]
Comment (chanh) [reply flashinfer/prefill.py]: Fixed — replaced the Python loop with `torch.repeat_interleave(valid_offsets[:-1], item_lens_b)`, fully vectorized with no GPU-CPU sync.

File: flashinfer/prefill.py [hunk 5]
```diff
@@ -1710,6 +1728,46 @@ def plan(
...
+            # Pad and concatenate
+            self._item_start_len = max_items_tokens
+            padded = []
+            for ist in item_start_list:
+                if len(ist) < max_items_tokens:
+                    ist = torch.cat([ist, torch.zeros(max_items_tokens - len(ist), dtype=torch.uint32, device=ist.device)])
+                padded.append(ist)
+            self._item_start_ptr = torch.cat(padded).to(item_offsets.device)
```
[the remark this answers is not in the document]
Comment (chanh) [reply flashinfer/prefill.py]: Fixed — `_precompute_item_start()` now takes `device` as a parameter, and both call sites pass `self.device`.

File: flashinfer/prefill.py [hunk 6]
```diff
@@ -1710,6 +1728,46 @@ def plan(
...
+            # Compute max_item_len from offsets
+            if max_item_len_ptr is None:
+                max_lens = []
+                for b in range(batch_size):
+                    offsets_b = item_offsets[b * item_offsets_len:(b + 1) * item_offsets_len]
+                    num_offsets = (offsets_b.to(torch.int64) > 0).sum().item() + 1
+                    valid_offsets = offsets_b[:num_offsets].long()
+                    diffs = valid_offsets[1:] - valid_offsets[:-1]
+                    max_lens.append(diffs.max().item())
```
[the remark this answers is not in the document]
Comment (chanh) [reply flashinfer/prefill.py]: Fixed — `max_item_len` computation is now also vectorized inside `_precompute_item_start()` using `(valid_offsets[1:] - valid_offsets[:-1]).max()`.

File: flashinfer/prefill.py [hunk 7]
```diff
@@ -1548,6 +1564,8 @@ def plan(
         token_pos_in_items_ptr: Optional[torch.Tensor] = None,
         token_pos_in_items_len: int = 0,
         max_item_len_ptr: Optional[torch.Tensor] = None,
+        item_offsets: Optional[torch.Tensor] = None,
+        item_offsets_len: int = 0,
```
[the remark this answers is not in the document]
Comment (chanh) [reply flashinfer/prefill.py:1767]: Good point. Currently MIS v2 is only wired through the FA2 and Hopper (SM90) code paths — the cudnn, trtllm-gen, cutlass, and FP8 branches do not read `_item_start_ptr`. I will add a validation check in `plan()` to reject `item_offsets` when the resolved backend does not support MIS v2. Will address in a follow-up commit.

File: flashinfer/prefill.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (chanh) [reply flashinfer/prefill.py]: Fixed — extracted `_precompute_item_start()` helper that takes `batch_size` from the caller, which is always `len(qo_indptr) - 1`. No more `len(prefix_len_ptr)` fallback.

File: flashinfer/prefill.py [hunk 8]
```diff
@@ -1710,6 +1728,46 @@ def plan(
...
+            # Compute max_item_len from offsets
+            if max_item_len_ptr is None:
+                max_lens = []
+                for b in range(batch_size):
+                    offsets_b = item_offsets[b * item_offsets_len:(b + 1) * item_offsets_len]
+                    num_offsets = (offsets_b.to(torch.int64) > 0).sum().item() + 1
+                    valid_offsets = offsets_b[:num_offsets].long()
+                    diffs = valid_offsets[1:] - valid_offsets[:-1]
+                    max_lens.append(diffs.max().item())
+                self._max_item_len_ptr = torch.tensor(max_lens, dtype=torch.uint16, device=item_offsets.device)
```
[the remark this answers is not in the document]
Comment (chanh) [reply flashinfer/prefill.py]: Acknowledged — this is a valid concern. MIS v2 currently allocates fresh `_item_start_ptr` on every `plan()` call, which is unsafe under CUDA graph capture. For now, MIS v2 is not intended for use with `use_cuda_graph=True`. I will add a guard to reject `item_offsets` when CUDA graphs are enabled, and address fixed-buffer support in a follow-up.

File: tests/attention/test_batch_prefill_kernels.py [hunk 9]
```diff
@@ -1022,6 +1022,316 @@ def create_2D_multi_item_mask_dense(
...
+    import torch
+
+    total_items_tokens = sum(item_lens)
+    # V2 has no delimiter tokens in the sequence
+    qo_len = total_items_tokens  # items region only (prefix is in KV cache)
+    kv_len = prefix_len + total_items_tokens
+
+    # Build item_offsets: CSR-style [0, len1, len1+len2, ...]
+    offsets = [0]
+    for l in item_lens:
```
[the remark this answers is not in the document]
Comment (chanh) [reply tests/attention/test_batch_prefill_kernels.py]: Fixed — renamed `l` to `item_len` in tests and `il` in the benchmark file.

File: flashinfer/prefill.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (chanh) [reply flashinfer/prefill.py:1767]: Yes, please go ahead and open a tracking issue for the backend validation guard. That would be helpful.

File: flashinfer/prefill.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (chanh) [reply flashinfer/prefill.py]: Good catch — will apply the guard in both wrapper classes. This can go in the same follow-up as the backend validation guard.

File: flashinfer/prefill.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (chanh) [reply flashinfer/prefill.py:1767]: Thanks for opening the tracking issue (#3138). I will self-assign it. This thread is resolved — the guard will be added in the follow-up.

[the remark this answers is not in the document]
Comment (chanh) [reply flashinfer/prefill.py:1767]: Addressed in this PR now — added `ValueError` guards in both `BatchPrefillWithPagedKVCacheWrapper.plan()` and `BatchPrefillWithRaggedKVCacheWrapper.plan()`:

1. **Backend guard**: rejects `item_offsets` when backend is `cudnn`, `trtllm-gen`, or `cutlass` (after backend resolution)
2. **CUDA graph guard**: rejects `item_offsets` when `use_cuda_graph=True` (paged wrapper only, since ragged does not support CUDA graphs)

Both guards fire before `_precompute_item_start` or at the earliest safe point, so no partial state mutations occur. Closes #3138.

File: flashinfer/prefill.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (chanh) [reply flashinfer/prefill.py]: Implemented in latest commit — added the CUDA graph guard in `BatchPrefillWithPagedKVCacheWrapper.plan()` before `_precompute_item_start` is called. The ragged wrapper does not support `use_cuda_graph` so no guard needed there.
